Add documentation on using virtual environments for testing in napari based on 2022-01 workshop by Talley Lambert

Nadalyn-CZI opened the pull request:
# Description
Adding content on virtual environments.
## Type of change
- [x] This change required adding documentation

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

Nadalyn-CZI committed:
Adding testing workshop documents and images.

Nadalyn-CZI committed:
Renamed Napari plugins to Readers and fixtures and added TOC.

Nadalyn-CZI committed:
Adding Talley's content.

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Developer-tools.md

Nadalyn-CZI committed:
Update More-than-just-best-practices.md

Nadalyn-CZI committed:
Update More-than-just-best-practices.md

Nadalyn-CZI committed:
Update Developer-tools.md

Nadalyn-CZI committed:
Update Virtual-environments.md

Nadalyn-CZI committed:
Update Virtual-environments.md

Nadalyn-CZI committed:
Update Virtual-environments.md

Nadalyn-CZI committed:
Update cover_page.md

Nadalyn-CZI committed:
Update cover_page.md

Nadalyn-CZI committed:
Update Virtual-environments.md

Nadalyn-CZI committed:
Update Virtual-environments.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Version-management.md

Nadalyn-CZI committed:
Update Version-management.md

Nadalyn-CZI committed:
Update Version-management.md

Nadalyn-CZI committed:
Update Developer-tools.md

Nadalyn-CZI committed:
Update More-than-just-best-practices.md

Nadalyn-CZI committed:
Update Survey.md

Nadalyn-CZI committed:
Update Version-management.md

Nadalyn-CZI committed:
Update Version-management.md

Nadalyn-CZI committed:
Update Survey.md

Nadalyn-CZI committed:
Update Survey.md

Nadalyn-CZI committed:
Update Virtual-environments.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Version-management.md

Nadalyn-CZI committed:
Update Developer-tools.md

Nadalyn-CZI committed:
Update More-than-just-best-practices.md

Nadalyn-CZI committed:
Update Deploying-your-plugin.md

Nadalyn-CZI committed:
Update Version-management.md

Czaki commented:
The build is failing. Did it pass when you built it locally?

Nadalyn-CZI commented:
No.

On Fri, Nov 18, 2022 at 3:14 AM Grzegorz Bokota ***@***.***>
wrote:

> The build is failing. Did it pass when you built it locally?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/napari/docs/pull/50#issuecomment-1319735309>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3CMNTFKDLUQ6G3VIXQXJBLWI5CFVANCNFSM6AAAAAASCVNCNQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

jni commented:
Thanks @Nadalyn-CZI! I don't have time for a thorough review right now — on quick glance these pages look great — but I just wanted to make a couple of notes:

- Ideally the PR title (and thus subsequent commit in the repo) should be descriptive to someone coming in without context. So my suggestion would be something like: "Add documentation on testing in napari based on 2022-01 workshop by Talley Lambert"
- spaces, special characters, and to a lesser extent capital letters are painful to deal with in the command line, which is why all of our filenames are typically all lower case and use dashes or underscores instead of whitespace wherever possible. Could you please rename files and folders to not have spaces or `&` or capitals?

Nadalyn-CZI commented:
Thank you.

On Sun, Nov 20, 2022 at 9:50 AM Juan Nunez-Iglesias <
***@***.***> wrote:

> Thanks @Nadalyn-CZI <https://github.com/Nadalyn-CZI>! I don't have time
> for a thorough review right now — on quick glance these pages look great —
> but I just wanted to make a couple of notes:
>
>    - Ideally the PR title (and thus subsequent commit in the repo) should
>    be descriptive to someone coming in without context. So my suggestion would
>    be something like: "Add documentation on testing in napari based on 2022-01
>    workshop by Talley Lambert"
>    - spaces, special characters, and to a lesser extent capital letters
>    are painful to deal with in the command line, which is why all of our
>    filenames are typically all lower case and use dashes or underscores
>    instead of whitespace wherever possible. Could you please rename files and
>    folders to not have spaces or & or capitals?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/napari/docs/pull/50#issuecomment-1321165723>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3CMNTGYHNKE5KAEVNR3M4LWJJCF3ANCNFSM6AAAAAASCVNCNQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Nadalyn-CZI committed:
Removing testing workshop docs from virtual environment workshop docs.

Nadalyn-CZI committed:
Merge branch 'Talley_workshop' of https://github.com/Nadalyn-CZI/docs into Talley_workshop

Nadalyn-CZI committed:
Update cover_page.md

Nadalyn-CZI committed:
Renamed files.

Nadalyn-CZI committed:
Merge branch 'Talley_workshop' of https://github.com/Nadalyn-CZI/docs into Talley_workshop

Nadalyn-CZI committed:
merge changes

Nadalyn-CZI committed:
Removed file name capital letters and made other changes to filenames.

Nadalyn-CZI commented:
Could we please get a final review on this?  I'd really like to get it published.  Requesting @tlambert03, @jni, @DragaDoncila, @chili-chiu.  Thanks!

Nadalyn-CZI commented:
@tlambert03 Would you please review these documents?

Nadalyn-CZI commented:
@jni @DragaDoncila please review.

tlambert03 commented:
Hi @Nadalyn-CZI,
I'm no longer working on napari, this is fine with me, but you might want approval from an active developer

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-1-virtual-environments.md:
```diff
@@ -0,0 +1,69 @@
+* ~1.5-hour lecture on getting set up with python and virtual environments:
+https://nic.med.harvard.edu/python/
+
+* Use miniconda or mini forge. Install it in the home directory.
```
I would turn the `miniconda` and `mini forge` links into hyperlinks i.e.

```suggestion
* Use [miniconda](https://docs.conda.io/en/latest/miniconda.html) or [mini forge](https://github.com/conda-forge/miniforge) (comes pre-configured with conda-forge). Install it in the home directory.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-1-virtual-environments.md:
```diff
@@ -0,0 +1,69 @@
+https://docs.conda.io/en/latest/miniconda.html  
+https://github.com/conda-forge/miniforge  (comes pre-configured with conda-forge) 
+
+* Make sure you have conda-forge in your config. The default anaconda channel has some very outdated packages, e.g. old version of Qt.   
```
```suggestion
* Make sure you have conda-forge in your config. The default anaconda channel has some very outdated packages, e.g. [old version of Qt](https://forum.image.sc/t/napari-issues-on-bigsur/52630/10).   
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-1-virtual-environments.md:
````diff
@@ -0,0 +1,69 @@
+
+* Make sure you have conda-forge in your config. The default anaconda channel has some very outdated packages, e.g. old version of Qt.   
+https://forum.image.sc/t/napari-issues-on-bigsur/52630/10  
+```python
````
```suggestion
```console
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-1-virtual-environments.md:
````diff
@@ -0,0 +1,69 @@
+
+To create an environment, use the following commands at the command prompt (terminal):
+
+```python
````
```suggestion
```console
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-1-virtual-environments.md:
````diff
@@ -0,0 +1,69 @@
+Consider your environment as disposable.
+If you are ever having weird problems, nuke your environment and start over using the following commands:  
+
+```python
````
```suggestion
```console
```

Pretty much all of these will be `console` unless it's writing a function or something

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,108 @@
+
+
+## PyPI and Anaconda
+Try to deploy to both! But for now, try to at least use pip. Currently, the built-in napari plugin installer and the napari-hub, will only look at PyPI. (This is subject to change in the future.) You can always provide your users with manual installation instructions (e.g. if you want to use conda).
```
We do actually support conda now (or we will very soon) so removing that caveat and slightly tweaking wording.

```suggestion
Try to deploy to both! But for now, try to at least use pip. You can always also provide your users with manual installation instructions (e.g. if you want to use conda or have specific dependencies).
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,108 @@
+
+**Note:** This goes for dependencies too! Check all your dependencies for wheel availability.  
+
+**Example:** See if a given package ships a wheel. https://pypi.org/project/napari/#files
```
I'm not really sure what this example is telling me to do/showing me?

Nadalyn-CZI replied:
I think it is showing you where to check for the wheel?  The example should probably be incorporated into the note.

Nadalyn-CZI replied:
I changed the text.

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,108 @@
+## Deploying plugins:
+
+### Automatically via GitHub actions
+This requires either: 
```
I would just simplify this into one sentence and then introduce the example workflow so:

This requires either running twine manually in a workflow or using a pre-made [GitHub action](https://github.com/pypa/gh-action-pypi-publish). Here is an example workflow that manually deploys using `twine` when tests pass and you push a tagged commit.

Nadalyn-CZI replied:
I'm not sure where that statement needs to go.  Does it need to go above ### PyPI?

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-3-version-management.md:
```diff
@@ -0,0 +1,48 @@
+* [Using a local script to edit files](#using-a-local-script-to-edit-files)  
+* [Manually](#manually)
+
+**Goal:** Make sure that you bump your version string everywhere it may appear, in unison, prior to publishing your package.  A package can be init.py, setup, cfg, etc.
```
```suggestion
**Goal:** Make sure that you bump your version string everywhere it may appear, in unison, prior to publishing your package.  A version number can be in `init.py`, `setup.cfg`, etc.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-3-version-management.md:
```diff
@@ -0,0 +1,48 @@
+   `# configure in pyproject.toml, then…`  
+   `$ git tag -a v0.1.0 -m v0.1.0`  
+
+  The next time you run <font color="red">python -m build</font>, either locally or in GitHub actions, your package version will be based on the latest git tag.
```
```suggestion
  The next time you run <font color="red">`python -m build`</font>, either locally or in GitHub actions, your package version will be based on the latest git tag.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-3-version-management.md:
```diff
@@ -0,0 +1,48 @@
+  The next time you run <font color="red">python -m build</font>, either locally or in GitHub actions, your package version will be based on the latest git tag.
+
+## Using a local script to edit files:  
+bump2version: [https://github.com/c4urself/bump2version](https://github.com/c4urself/bump2version)
```
```suggestion
One tool for doing this is [bump2version](https://github.com/c4urself/bump2version). For example:
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+
+### Linting tools   
+These _check_ your code.  
+* flake8 (https://flake8.pycqa.org/) - checks various code style conventions, unused variables, line spacings, etc…  
```
```suggestion
* [flake8](https://flake8.pycqa.org/) - checks various code style conventions, unused variables, line spacings, etc…  
```

All the links below should be updated in this manner for readability

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-4-developer-tools.md:
````diff
@@ -0,0 +1,76 @@
+
+### Pre-commit tools
+* pre-commit (https://pre-commit.com/), runs all your checks each time you run git commit, preventing bad code from ever getting checked in.  
+```console     $ pip install pre-commit
````
Need to put that `$ pip install pre-commit` bit of code on the next line.

Nadalyn-CZI replied:
Of course!  Thanks! Good eye!

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+    - It (currently) checks that your plugin is installable, and performs a few sanity checks about Qt backends and dock widgets.  
+`     uses: tlambert03/napari-plugin-action@main  `  
+`     with:     `     
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`package_name:  <your-package-name>`     
```
This whole bit should be in a yaml code block. Also should remove the `&nbsp`

Nadalyn-CZI replied:
I think I was trying to make this look like Talley's slide.  I couldn't get the spaces to stay in unless I used &nbsp.

Nadalyn-CZI replied:
I think I figured out where the yaml code block should start.

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-5-more-than-just-best-practices.md:
```diff
@@ -0,0 +1,44 @@
+# Article 5: More than just best practices  
```
TBH I think we could remove this whole article in favour of the already documented and linked best practices doc.

Nadalyn-CZI replied:
It is repetitive. I'll make sure it doesn't say anything that is not already in Best Practices.  If it doesn't, I'll remove it.

DragaDoncila commented on docs/plugins/virtual_environment_docs/article-6-survey.md:
```diff
@@ -0,0 +1,118 @@
+# Tools and tips when building napari plugins: Survey 
```
I'm not sure how useful this Q&A will be outside of the context of the workshop. I would vote to omit this personally. I believe everything here is covered in other docs. Curious what others think.

Nadalyn-CZI replied:
It seemed like it would be useful to beginners.

DragaDoncila replied:
Yeah there's a lot of useful info here. First thing I guess is I would maybe call it a Q&A rather than a survey? And I would make the questions either their own headings or dropdown admonitions so that the page is easier to browse at a glance. Right now I feel like the questions get lost in the doc.

Nadalyn-CZI replied:
Unless you have a strong objection, I'm going to leave this as is.

DragaDoncila replied:
@Nadalyn-CZI I think the formatting here is important, so it will need to be fixed before merge. Let me know if the suggestions are unclear.

DragaDoncila reviewed (commented):
Hi Nadalyn, have left some comments and general items listed below. Let me know if anything is unclear!

- We should rename all images to lowercase.
- We should delete from this branch any images that are already uploaded in #35.
- Just as with #35 I think it would be great if we could link to this from the existing [test and deploy](https://napari.org/stable/plugins/test_deploy.html#) doc or maybe on the Plugins index page?

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-1-virtual-environments.md

Added links to miniconda and miniforge.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-1-virtual-environments.md

Issues with default anaconda channel.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-1-virtual-environments.md

Changed from python formatting to console formatting.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-1-virtual-environments.md

Changed from ```python formatting to ```console formatting.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-1-virtual-environments.md

Changed from ```python to ```console.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-2-deploying-your-plugin.md

Additional wording about manual instructions.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-3-version-management.md

Clarification.  Use 'version number' instead of 'package.'

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-3-version-management.md

Correcting formatting of console type text.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-3-version-management.md

Clarifying wording.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/article-4-developer-tools.md

Update links for readability.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update article-2-deploying-your-plugin.md

Wording clarification.

Nadalyn-CZI committed:
Update article-4-developer-tools.md

Moved command to its own line.

Nadalyn-CZI committed:
Corrected formatting and some wording.

Nadalyn-CZI committed:
Update article-1-virtual-environments.md

Removed More than just Best Practices artcle.

Nadalyn-CZI committed:
Update article-2-deploying-your-plugin.md

Removed More than just Best Practices article.

Nadalyn-CZI committed:
Update article-3-version-management.md

Removed More than just Best Practices article.

Nadalyn-CZI committed:
Update article-4-developer-tools.md

Removed More than just Best Practices article.

Nadalyn-CZI committed:
Delete article-5-more-than-just-best-practices.md

Removed.  This is redundant information.

Nadalyn-CZI committed:
Update article-4-developer-tools.md

Replaced More than just Best Practices with Survey article.

Nadalyn-CZI committed:
Update index.md

Removed More than just Best Practices article.

Nadalyn-CZI committed:
Update article-2-deploying-your-plugin.md

Clarification of wording.

Nadalyn-CZI committed:
Clean up directory.

Nadalyn-CZI committed:
Clean up directory.

Nadalyn-CZI committed:
Changed titles of all articles.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+* 2: [Deploying your plugin](./2-deploying-your-plugin.md)    
+* 3: [Version management](./3-version-management.md)   
+* 4: This guide   
+* 5: [Survey](./5-Survey.md) 
```
```suggestion
* 5: [Survey](./5-survey.md) 
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+* 2: [Deploying your plugin](./2-deploying-your-plugin.md)  
+* 3: [Version management](./3-version-management.md)     
+* 4: [Developer tools](./4-developer-tools.md)   
+* 5: [Survey](./5-Survey.md)   
```
```suggestion
* 5: [Survey](./5-survey.md)   
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+* [Creating environments](#creating-environments)  
+* [Removing environments](#removing-environments)
+
+</br>
```
I think here we should remove the manual whitespace added, and expand a little bit on 
1. what a virtual environment is and
2. why they are super important

At the top we say we're going to 'explain the value of using virtual environments' and I don't think this guide currently does that.

Nadalyn-CZI replied:
I added some content here.  See if you think it is enough and is accurate.

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+https://nic.med.harvard.edu/python/
+
+* Use [miniconda](https://docs.conda.io/en/latest/miniconda.html) or [mini forge](https://github.com/conda-forge/miniforge) (comes pre-configured with conda-forge). Install it in the home directory.
+https://docs.conda.io/en/latest/miniconda.html  
```
We should remove this link and the one below since we've added hyperlinks above.
```suggestion

```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+* Use [miniconda](https://docs.conda.io/en/latest/miniconda.html) or [mini forge](https://github.com/conda-forge/miniforge) (comes pre-configured with conda-forge). Install it in the home directory.
+https://docs.conda.io/en/latest/miniconda.html  
+https://github.com/conda-forge/miniforge  (comes pre-configured with conda-forge) 
```
```suggestion
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+https://github.com/conda-forge/miniforge  (comes pre-configured with conda-forge) 
+
+* Make sure you have conda-forge in your config. The default anaconda channel has some very outdated packages, e.g. [old version of Qt](https://forum.image.sc/t/napari-issues-on-bigsur/52630/10).   
+https://forum.image.sc/t/napari-issues-on-bigsur/52630/10  
```
```suggestion
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
````diff
@@ -0,0 +1,68 @@
+
+* Make sure you have conda-forge in your config. The default anaconda channel has some very outdated packages, e.g. [old version of Qt](https://forum.image.sc/t/napari-issues-on-bigsur/52630/10).   
+https://forum.image.sc/t/napari-issues-on-bigsur/52630/10  
+```console
````
Should add a sentence before this code snippet explaining briefly what it does:

These commands add the `conda-forge` channel to the `conda` config, so that packages in `conda-forge` are visible to the `conda` installer. Setting `channel_priority` to `strict` ensures packages in high priority channels are always installed over packages of the same name in lower priority channels. See [this guide](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html) for more details

Nadalyn-CZI replied:
I think I addressed this.

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+## Resources
+Some resources you might find helpful: 
+* ~1.5-hour lecture on getting set up with python and virtual environments:
+https://nic.med.harvard.edu/python/
```
This link should be a hyperlink on the word `lecture`.

Nadalyn-CZI replied:
This is the dead link.  If you follow it you get a 404 error.  I changed it.

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+This guide explains the value of using virtual environments and how to create and remove them. 
+
+## Other topics in this series:  
```
This section + `This guide covers` take a lot of room before we actually get to the content. I wonder if the `Other topics in this series` section should go at the end? Especially because these guides are not sequential, like the ones in the testing workshop. I think that's also more consistent with other websites I've seen, where when you get to the bottom it shows you other things you could look at.

Nadalyn-CZI replied:
So, we should delete the statement that says, "The next topic in this series is..."? I moved the descriptions.  We have numbered them just because that is the order they were presented in.  You're right, they don't necessarily build on each other.  Should we remove the numbers too?

The last section is now: 

## Other topics in this series:  

* 2: [Deploying your plugin](./2-deploying-your-plugin.md)  
* 3: [Version management](./3-version-management.md)     
* 4: [Developer tools](./4-developer-tools.md)   
* 5: [Survey](./5-survey.md)

DragaDoncila replied:
@Nadalyn-CZI yeah agreed, I think we can get rid of the numbers here.

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+</br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. . . are super important!  
+</br>  
+
+## Resources
```
I think the key part of this section is the link to installing `miniconda` or `miniforge`, and setting up the `conda-forge` channel.

So I think maybe we should call it `Setting Up` or something similar. We should remove the bullet points and turn it into the following narrative:

```
## Setting Up

Sentences from above about what a virtual environment is and why it's important.

Sentence about how conda is one of the most popular ones.

Links to install miniconda or mini forge

Instructions to add conda-forge to config

Finally, link to lecture for who wants more info on this.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
````diff
@@ -0,0 +1,68 @@
+    $ pip/conda install <whatever> 
+```
+
+Virtual environments  
````
I think this bit should go into the `Removing environments` section. I also think we should remove the whitespace. I do like the stylistic effect, but we don't really use this anywhere else on the website. If we want to emphasize this statement perhaps there's a theme element we can use? Maybe a [line block](https://pydata-sphinx-theme.readthedocs.io/en/stable/examples/kitchen-sink/blocks.html#line-blocks)?

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,107 @@
+
+
+## PyPI and Anaconda
+Try to deploy to both! But for now, try to at least use pip. You can always also provide your users with manual installation instructions (e.g. if you want to use conda or have specific dependencies).
```
Would add just a short sentence here about the two:

```suggestion
PyPI and Anaconda are two options for how you distribute your package and allow your users to more easily find and install it.  Try to deploy to both! But for now, try to at least use PyPI. You can always also provide your users with manual installation instructions (e.g. if you want them to use `conda` or have specific dependencies).
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,107 @@
+
+**Note:** This goes for dependencies too! Check all your dependencies for wheel availability.  
+
+### PyPI
```
This section should be moved to `Deploying plugins` and it should be called `Manually via twine`.

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
````diff
@@ -0,0 +1,107 @@
+### PyPI
+**build** (package builder): https://pypa-build.readthedocs.io/en/latest/  
+**twine** (PyPI Client): https://twine.readthedocs.io/en/latest/  
+```python
````
```suggestion
```console
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,107 @@
+**Note:** This goes for dependencies too! Check all your dependencies for wheel availability.  
+
+### PyPI
+**build** (package builder): https://pypa-build.readthedocs.io/en/latest/  
```
```suggestion
**[build](https://pypa-build.readthedocs.io/en/latest/ )** : package builder that bundles your source code into `sdist` or `wheel` distributions
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,107 @@
+
+### PyPI
+**build** (package builder): https://pypa-build.readthedocs.io/en/latest/  
+**twine** (PyPI Client): https://twine.readthedocs.io/en/latest/  
```
```suggestion
**[twine](https://twine.readthedocs.io/en/latest/ )**: client you can use to upload your distribution to PyPI. Note that you will need to set up a PyPI account and authenticate yourself when uploading. See [this great guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/) for a detailed tutorial to building & sharing your first Python packages.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
````diff
@@ -0,0 +1,107 @@
+    $ twine upload dist/*  
+
+```
+**Note:** `python -m build` is the modern alternative to setuptools `python setup.py sdist bdist_wheel`. (It calls setuptools behind the scenes.)
````
```suggestion
**Note:** `python -m build` is the modern alternative to `setuptools`' `python setup.py sdist bdist_wheel`. It calls `setuptools` behind the scenes.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,107 @@
+
+### Automatically via GitHub actions
+This requires either: 
+* Running twine manually (e.g: your-repo/.github/workflows/deploy.yaml)  
```
```suggestion
* Running `twine` as above in a workflow after setting up Python and installing it
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,107 @@
+This requires either: 
+* Running twine manually (e.g: your-repo/.github/workflows/deploy.yaml)  
+or   
+* Using a pre-made GitHub action: https://github.com/pypa/gh-action-pypi-publish   
```
```suggestion
* Using a pre-made [GitHub action](https://github.com/pypa/gh-action-pypi-publish)
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
````diff
@@ -0,0 +1,107 @@
+
+Here is an example workflow that manually deploys using `twine` when tests pass and you push a tagged commit.
+
+```yaml
````
```suggestion
```yaml
#your-repo/.github/workflows/deploy.yaml
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
````diff
@@ -0,0 +1,107 @@
+
+```
+
+  **Note:** Gate this action on some criterion, like a git tag: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_Idif
````
This note makes more sense just after line 62 where we first mention a tagged commit. Also should be a hyperlink.

```suggestion
  **Note:** Gate this action on some criterion, like a [git tag](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_Idif).
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,47 @@
+* [Using a local script to edit files](#using-a-local-script-to-edit-files)  
+* [Manually](#manually)
+
+**Goal:** Make sure that you bump your version string everywhere it may appear, in unison, prior to publishing your package.  A version number can be in `init.py`, `setup.cfg`, etc.
```
```suggestion
Your goal is to make sure that you bump your version string everywhere it may appear, in unison, prior to publishing your package.  A version number can be in `init.py`, `setup.cfg`, etc.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,47 @@
+
+**Goal:** Make sure that you bump your version string everywhere it may appear, in unison, prior to publishing your package.  A version number can be in `init.py`, `setup.cfg`, etc.
+
+In increasing order of work, but decreasing order of magic, the methods of bummping your version string are listed below. 
```
```suggestion
In increasing order of work, but decreasing order of magic, the methods of bumping your version string are listed below. 
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,47 @@
+In increasing order of work, but decreasing order of magic, the methods of bummping your version string are listed below. 
+
+## Using git tags:  
+setuptools_scm: [https://github.com/pypa/setuptools_scm](https://github.com/pypa/setuptools_scm)
```
```suggestion
You can use [setuptools_scm](https://github.com/pypa/setuptools_scm) to automatically generate version numbers for your package based on tagged commits.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,47 @@
+   `# configure in pyproject.toml, then…`  
+   `$ git tag -a v0.1.0 -m v0.1.0`  
+
+  The next time you run <font color="red">`python -m build`</font>, either locally or in GitHub actions, your package version will be based on the latest git tag.
```
The font color specification isn't working as I think the backtick formatting is taking precedence. I would just remove it.

```suggestion
  The next time you run `python -m build`, either locally or in GitHub actions, your package version will be based on the latest git tag.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
````diff
@@ -0,0 +1,47 @@
+```console
+   $ pip install bump2version  
+   # then configure all the places you use your version   
+   $ bump2version -$current-version 0.5.1 minor  
````
```suggestion
   $ bump2version --current-version 0.5.1 minor  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
````diff
@@ -0,0 +1,47 @@
+```   
+
+## Manually
+This is ***not*** recommended, you *will* eventually make mistakes and have mismatched version/metadata somewhere.  
````
It's a bit of a red herring to have the header if we don't actually explain how to do it manually lol.

```suggestion
Updating the version number manually involves going through everywhere your version is declared and changing the version number before building your distribution. This is ***not*** recommended, you *will* eventually make mistakes and have mismatched version/metadata somewhere. In some cases this will lead to your build process failing, but it can fail silently too.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,47 @@
+## Manually
+This is ***not*** recommended, you *will* eventually make mistakes and have mismatched version/metadata somewhere.  
+  
+  ### *The "best" versioning and deployment workflow is the one you will actually use!*  
```
I would turn this sentence and the one below into a `tip` admonition and remove the additional styling.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+    - [Linting tools](#linting-tools)  
+    - [Formatting tools](#formatting-tools)  
+    - [Pre-commit tools](#pre-commit-tools)  
+* [Napari-plugin-specific-developer-tools](#napari-plugin-specific-developer-tools)
```
```suggestion
* [napari plugin-specific tools](#napari-plugin-specific-tools)
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+    - Runs all your pre-commit hooks on CI (Continuous Integration).
+    - Useful even if contributors don't install and run your pre-commit hooks locally before they open a PR.  
+  
+## Napari-plugin-specific developer tools  
```
```suggestion
## Napari plugin-specific tools  
```

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+### Linting tools   
+These _check_ your code.  
+* [flake8](https://flake8.pycqa.org/) - checks various code style conventions, unused variables, line spacings, etc…  
+* mypy (https://github.com/python/mypy)  
```
```suggestion
* [mypy](https://github.com/python/mypy)  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+
+### Formatting tools 
+These _auto-modify_ your code.  
+* black (https://github.com/psf/black)  
```
```suggestion
* [black](https://github.com/psf/black)  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+These _auto-modify_ your code.  
+* black (https://github.com/psf/black)  
+  Forces code to follow specific style, indentations, etc...  
+* autoflake (https://github.com/myint/autoflake)  
```
```suggestion
* [autoflake](https://github.com/myint/autoflake)  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+  Forces code to follow specific style, indentations, etc...  
+* autoflake (https://github.com/myint/autoflake)  
+  Auto-fixes some flake8 failures.  
+* isort (https://github.com/PyCQA/isort)  
```
```suggestion
* [isort](https://github.com/PyCQA/isort)  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+  Auto-fixes some flake8 failures.  
+* isort (https://github.com/PyCQA/isort)  
+  Auto-sorts and formats your imports.
+* setup-cfg-fmt (https://github.com/asottile/setup-cfg-fmt)  
```
```suggestion
* [setup-cfg-fmt](https://github.com/asottile/setup-cfg-fmt)  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+  Sorts and enforces conventions in setup.cfg.  
+
+### Pre-commit tools
+* pre-commit (https://pre-commit.com/), runs all your checks each time you run git commit, preventing bad code from ever getting checked in.  
```
```suggestion
* [pre-commit](https://pre-commit.com/), runs all your checks each time you run git commit, preventing bad code from ever getting checked in.  
```

Nadalyn-CZI replied:
done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+     $ pre-commit install  
+     # then configure in .pre-commit-config.yaml  
+     # (optionally) Run hooks on demand  
+     $ pre-commit run -$all-files  
```
```suggestion
     $ pre-commit run --all-files  
```

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
````diff
@@ -0,0 +1,76 @@
+     $ pre-commit run -$all-files  
+```  
+
+* pre-commit-ci (https://pre-commit.ci/)
````
```suggestion
* [pre-commit-ci](https://pre-commit.ci/)
```

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+  
+## Napari-plugin-specific developer tools  
+
+* https://github.com/tlambert03/napari-plugin-checks 
```
```suggestion
* [Static plugin checks](https://github.com/tlambert03/napari-plugin-checks)
```

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
````diff
@@ -0,0 +1,76 @@
+    hooks: id: napari-plugin-checks  
+```     
+
+* https://github.com/tlambert03/napari-plugin-action  (work in progress)
````
```suggestion
* [Plugin check GitHub action](https://github.com/tlambert03/napari-plugin-action)  (work in progress)
```

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
````diff
@@ -0,0 +1,76 @@
+```     
+
+* https://github.com/tlambert03/napari-plugin-action  (work in progress)
+    - This is a _github-action_. 
````
```suggestion
```

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/4-developer-tools.md:
```diff
@@ -0,0 +1,76 @@
+    - This is a _github-action_. 
+    - It is intended to be added to your GitHub workflow.
+    - It (currently) checks that your plugin is installable, and performs a few sanity checks about Qt backends and dock widgets.  
+`     uses: tlambert03/napari-plugin-action@main`  
```
This should be in one yaml block

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,117 @@
+# Tools and tips when building napari plugins: Survey 
```
```suggestion
# 5: Survey
```

Nadalyn-CZI replied:
Done.

DragaDoncila reviewed (changes requested):
Hi @Nadalyn-CZI finally have had time to do a detailed review of this one. I tried to leave direct suggestions where I could. Let me know if something is unclear 😊 

We also need to add these docs to the TOC, like we did for PR #35

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/1-virtual-environments.md

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update docs/plugins/virtual_environment_docs/4-developer-tools.md

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Apply suggestions from code review

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Removing numbering of articles.

Nadalyn-CZI committed:
Removed numbering.

Nadalyn-CZI committed:
Removing numbering from articles in the workshop.

Nadalyn-CZI committed:
Removing numbering.

Nadalyn-CZI committed:
Adding content about conda.

Nadalyn-CZI committed:
Added content.

Added content about conda and the importance of virtual environments.

Nadalyn-CZI committed:
Update 1-virtual-environments.md

Nadalyn-CZI committed:
Clarification

Moved twine info to Deploying plugins.

Nadalyn-CZI committed:
Update 2-deploying-your-plugin.md

Nadalyn-CZI committed:
Clarified tips.

Nadalyn-CZI committed:
Various updates to formats & some clarifications.

Nadalyn-CZI reviewed (commented):
I believe I have made all changes.   @DragaDoncila or @psobolewskiPhD , please re-review.

Nadalyn-CZI committed:
Merge branch 'main' into Talley_workshop

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+## Overview
+A virtual environment is an isolated collection of packages, settings, and an associated Python interpreter, that allows multiple different collections to exist on the same system. They are created on top of an existing Python installation, known as the virtual environment's “base” python, and may optionally be isolated from the packages in the base environment, so only those explicitly installed ...
+
+ More information on why virtual environments are created and how they can help you can be found on the [python website](https://docs.python.org/3/library/venv.html#creating-virtual-environments) and at [hackmd.io](https://hackmd.io/@talley/SJB_lObBi#What-is-a-virtual-environment). 
```
`hackmd.io` is just the document hosting website, so I'm changing this hyperlink to be a bit more specific.

```suggestion
 More information on why virtual environments are created and how they can help you can be found on the [python website](https://docs.python.org/3/library/venv.html#creating-virtual-environments) and in [this introductory workshop](https://hackmd.io/@talley/SJB_lObBi#What-is-a-virtual-environment). 
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+ More information on why virtual environments are created and how they can help you can be found on the [python website](https://docs.python.org/3/library/venv.html#creating-virtual-environments) and at [hackmd.io](https://hackmd.io/@talley/SJB_lObBi#What-is-a-virtual-environment). 
+
+Virtual environments are super important! They allow you to isolate your project from other python projects. They allow you to experiment with various packages and versions without fear of breaking your entire system (and needing to reinstall everything). As you install packages over time, you will inevitably install something that doesn’t “play well” with something else that is already installed ...
```
```suggestion
Virtual environments are super important! They allow you to isolate your project from other Python projects. They allow you to experiment with various packages and versions without fear of breaking your entire system (and needing to reinstall everything). As you install packages over time, you will inevitably install something that doesn’t “play well” with something else that is already installed. In some cases this can be hard to recover from. With virtual environments, you can just create a fresh environment and start again – without needing to do major surgery on your system.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+There are several tools available for creating and managing virtual environments.  One of the most popular, comprehensive tools is Conda. Wikipedia explains that [Conda is an open-source, cross-platform, language-agnostic package manager and environment management system.](https://en.wikipedia.org/wiki/Conda_(package_manager)) It was originally developed to solve difficult package management chal ...
+
+For more information on the value of virtual environments, see [hackmd.io](https://hackmd.io/@talley/SJB_lObBi#What-is-a-virtual-environment). 
```
We already have this workshop linked above so I think we should remove this sentence.

```suggestion
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+Virtual environments are super important! They allow you to isolate your project from other python projects. They allow you to experiment with various packages and versions without fear of breaking your entire system (and needing to reinstall everything). As you install packages over time, you will inevitably install something that doesn’t “play well” with something else that is already installed ...
+
+There are several tools available for creating and managing virtual environments.  One of the most popular, comprehensive tools is Conda. Wikipedia explains that [Conda is an open-source, cross-platform, language-agnostic package manager and environment management system.](https://en.wikipedia.org/wiki/Conda_(package_manager)) It was originally developed to solve difficult package management chal ...
```
```suggestion
There are several tools available for creating and managing virtual environments.  One of the most popular, comprehensive tools is Conda. Wikipedia explains that [Conda is an open-source, cross-platform, language-agnostic package manager and environment management system.](https://en.wikipedia.org/wiki/Conda_(package_manager)) It was originally developed to solve difficult package management challenges faced by Python data scientists. The Conda package and environment manager is included in all versions of **Anaconda**, **Miniconda**, and **Anaconda Repository**. 
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+There are several tools available for creating and managing virtual environments.  One of the most popular, comprehensive tools is Conda. Wikipedia explains that [Conda is an open-source, cross-platform, language-agnostic package manager and environment management system.](https://en.wikipedia.org/wiki/Conda_(package_manager)) It was originally developed to solve difficult package management chal ...
+
+For more information on the value of virtual environments, see [hackmd.io](https://hackmd.io/@talley/SJB_lObBi#What-is-a-virtual-environment). 
+
```
Would be good to add this as a heading so users who skim the overview don't miss the install step.

```suggestion
## Install and Config
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+Adding the conda-forge channel to the conda config makes packages in conda-forge visible to the conda installer. Setting channel_priority to strict ensures packages in high priority channels are always installed over packages of the same name in lower priority channels. See [this guide](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html) for more details. 
+
+Make sure conda-forge is in your config by using the following commands:
```
```suggestion
Make sure the `conda-forge` channel is in your config by using the following commands:
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+Install [miniconda](https://docs.conda.io/en/latest/miniconda.html) or [mini forge](https://github.com/conda-forge/miniforge) (comes pre-configured with conda-forge) in the home directory.
+
+Adding the conda-forge channel to the conda config makes packages in conda-forge visible to the conda installer. Setting channel_priority to strict ensures packages in high priority channels are always installed over packages of the same name in lower priority channels. See [this guide](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html) for more details. 
```
```suggestion
Adding the `conda-forge` channel to the conda config makes packages in `conda-forge` visible to the conda installer. Setting `channel_priority` to strict ensures packages in high priority channels are always installed over packages of the same name in lower priority channels. See [this guide](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html) for more details. 
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/1-virtual-environments.md:
```diff
@@ -0,0 +1,68 @@
+
+For more information on the value of virtual environments, see [hackmd.io](https://hackmd.io/@talley/SJB_lObBi#What-is-a-virtual-environment). 
+
+Install [miniconda](https://docs.conda.io/en/latest/miniconda.html) or [mini forge](https://github.com/conda-forge/miniforge) (comes pre-configured with conda-forge) in the home directory.
```
```suggestion
Install [miniconda](https://docs.conda.io/en/latest/miniconda.html) or [mini forge](https://github.com/conda-forge/miniforge) (comes pre-configured with `conda-forge`) in the home directory.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+* [PyPI and Anaconda](#pypi-and-anaconda)
+    - [Package formats (PyPI)](#package-formats-pypi)
+    - [PyPI](#pypi)
+* [Deploying plugins](#deploying-plugins)
```
```suggestion
* [Deploying plugins](#deploying-plugins)
    - [Manually via twine](#manually-via-twine)
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+
+This guide explains some of the techniques you can use to deploy your plugin. 
+
+## This guide covers:   
```
Hmm this structure is a little bit confusing. We only really talk about Anaconda in the very last bit, so I would make that its own heading, and focus on PyPI in the first two sections. So this would be my suggested structure:

```
* PyPI package formats
    - Overview
    - Building your package
* Deploying to PyPI
    - Manually via twine
    - Automatically via GitHub actions
* Deploying to Anaconda
```

Nadalyn-CZI replied:
Good idea. I will change this.

Nadalyn-CZI replied:
Ok, @DragaDoncila, please see what you think now.  It is not exactly like you suggested but I think it makes sense.

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+**Note:** This goes for dependencies too! Check all your dependencies for wheel availability.  
+
+### PyPI
+**[build](https://pypa-build.readthedocs.io/en/latest/ )** : package builder that bundles your source code into `sdist` or `wheel` distributions
```
```suggestion
**[build](https://pypa-build.readthedocs.io/en/latest/ )** is the recommended package builder that bundles your source code into `sdist` or `wheel` distributions. Install `build` into your local environment, and then run it at the root of your package to build your package:
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+
+### PyPI
+**[build](https://pypa-build.readthedocs.io/en/latest/ )** : package builder that bundles your source code into `sdist` or `wheel` distributions
+
```
And the code block for the comment above...:

```console
pip install build
python -m build .
```

Nadalyn-CZI replied:
I thought this was covered in ### Automatically via GitHub actions?

DragaDoncila replied:
Building is slightly separate to deploying. You need to build to deploy, but I think we should have those two commands here to demonstrate the last sentence: _"Install `build` into your local environment and then run it at the root of your package to build your package."_

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+## Deploying plugins:
+
+### Manually via **twine**.  
+[twine](https://twine.readthedocs.io/en/latest/ )**: client you can use to upload your distribution to PyPI. Note that you will need to set up a PyPI account and authenticate yourself when uploading. See [this great guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/) for a detailed tutorial to building and sharing your first Python packages.  
```
```suggestion
[twine](https://twine.readthedocs.io/en/latest/ ) is a command line client you can use to upload your distribution to PyPI. Note that you will need to set up a PyPI account and authenticate yourself when uploading. See [this great guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/) for a detailed tutorial to building and sharing your first Python packages.  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
````diff
@@ -0,0 +1,109 @@
+
+```
+
+  **Note:** Gate this action on some criterion, like a [git tag](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_Idif).
````
```suggestion
  **Note:** Gate this action on some criterion, e.g. a git tag as above, or [some other criterion](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+  **Note:** Gate this action on some criterion, like a [git tag](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_Idif).
+
+### Anaconda cloud via conda-forge
+https://conda-forge.org/docs/maintainer/adding_pkgs.html
```
```suggestion
This is only a brief guide to deploying to `conda-forge`. More information can be found in the [conda-forge docs](https://conda-forge.org/docs/maintainer/adding_pkgs.html).
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+
+Once your recipe is approved and merged, the rest happens *automagically*, and your package will appear on the anaconda cloud.
+
+This is **far** easier if you already have an sdist published to PyPI.
```
```suggestion
This is **far** easier if you already have an `sdist` published to PyPI.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
````diff
@@ -0,0 +1,44 @@
+One tool for doing this is [bump2version](https://github.com/c4urself/bump2version). For example:
+```console
+   $ pip install bump2version  
+   # then configure all the places you use your version   
````
```suggestion
   # configure all the places you use your version, then, to update:
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,44 @@
+## Manually
+Updating the version number manually involves going through everywhere your version is declared and changing the version number before building your distribution. This is ***not*** recommended, you *will* eventually make mistakes and have mismatched version/metadata somewhere. In some cases this will lead to your build process failing, but it can fail silently too.
+  
+  ### Tips: 
```
I think this should be an admonition to draw more attention.

Nadalyn-CZI replied:
I think you have to have an extension to use admonition. I could not figure out how to get the nice ones with the colored bar across the top like we have in our other documents.  I changed the level of the heading to make it stand out more.

Nadalyn-CZI replied:
Please let me know what you think.

melissawm replied:
The admonition should be added like so:

````
```{admoniton} Tips
  * The "best" versioning and deployment workflow is the one you will actually use!  
  * Get comfortable with at least one workflow for versioning and deploying your package *otherwise, you won't do it.*
```
````

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,44 @@
+* [Virtual environments](./1-virtual-environments)  
+* [Deploying your plugin](./2-deploying-your-plugin.md)    
+* [Developer tools](./4-developer-tools.md)   
+* [Survey](./5-Survey.md) 
```
```suggestion
* [Survey](./5-survey.md) 
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+* [Virtual environments](./1-virtual-environments.md)   
+* [Version management](./3-version-management.md)     
+* [Developer tools](./4-developer-tools.md)   
+* [Survey/Q&A](./5-Survey.md)   
```
```suggestion
* [Survey/Q&A](./5-survey.md)   
```

DragaDoncila reviewed (commented):
This is really shaping up @Nadalyn-CZI! Left a few more comments. Other outstanding issues:

- I still think the survey page should be formatted a bit more clearly, with questions as headings or maybe dropdowns so it's easier to browse the page. Currently the structure at the top and sidebar don't really make it clear to the user what each question/answer is about:
<img width="1032" alt="image" src="https://user-images.githubusercontent.com/17995243/220764327-db4ffce7-c253-41dd-9c19-ea110cef43e6.png">

- We still need to add this to the TOC. I'm not sure where it would be best suited. Some of this stuff is relevant outside of plugins, e.g. the virtual environment doc, but some of it is very specific. @melissawm I'm curious what you think.

Nadalyn-CZI committed:
Update index.md

Corrected filename of survey file.

Nadalyn-CZI committed:
Apply suggestions from Draga's code review

Filename corrections and editorial changes.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

psobolewskiPhD committed:
Merge branch 'main' into Talley_workshop

Nadalyn-CZI committed:
Update 2-deploying-your-plugin.md

Nadalyn-CZI committed:
Update 3-version-management.md

Nadalyn-CZI committed:
Update 3-version-management.md

melissawm commented:
> * We still need to add this to the TOC. I'm not sure where it would be best suited. Some of this stuff is relevant outside of plugins, e.g. the virtual environment doc, but some of it is very specific. @melissawm I'm curious what you think.

To me, this still reads as a workshop, so I would add it to the napari Workshop section.

DragaDoncila committed:
Merge branch 'main' into Talley_workshop

Nadalyn-CZI committed:
Adding virtual environments docs.

melissawm commented:
Ok - after talking to Nadalyn we figured this would fit the Plugins documentation better, so we put it as the last item in the plugin ToC:

https://output.circle-artifacts.com/output/job/18de8a46-7dcb-4fb2-9fbb-2f51254abba0/artifacts/0/docs/_build/plugins/index.html

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,109 @@
+PyPI and Anaconda are two options for how you distribute your package and allow your users to more easily find and install it.  Try to deploy to both! But for now, try to at least use PyPI. You can always also provide your users with manual installation instructions (e.g. if you want them to use `conda` or have specific dependencies).
+
+### Building your package 
+**sdist** means source distribution. An **sdist** includes all of the files that are required to *build* your package. An **sdist** may require specific additional software (e.g. compilers) to actually build.
```
```suggestion
`sdist` means source distribution. An `sdist` includes all of the files that are required to *build* your package. An `sdist` may require specific additional software (e.g. compilers) to actually build.
```

I think we should swap all bolded references to `sdist` and `wheel` with backticks.

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
```diff
@@ -0,0 +1,44 @@
+## Using git tags:  
+You can use [setuptools_scm](https://github.com/pypa/setuptools_scm) to automatically generate version numbers for your package based on tagged commits.
+
+   `# configure in pyproject.toml, then…`  
```
These two lines should be a codeblock with triple backticks.

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,115 @@
+    - [npe2](#npe2)  
+    - [Widget development](#widget-development)
+  
+## Questions
```
I think we should get rid of `##Questions` and `##Topics` as headings. The distinction isn't really there.

Nadalyn-CZI replied:
Please see what you think now.

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,115 @@
+  
+## Questions
+
+### Testing plugins    
```
These headings should just be the questions themselves. This would allow them to show in the sidebar, once we bump these headings to `##` after removing the `Questions` and `Topics`.

Nadalyn-CZI replied:
Please see what you think now.

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,115 @@
+Question: What are the best practices to test a plugin with multiple sequential steps?  
+e.g. Is it ok to rely on the "results" of a test to run the next test or should they all be fully independent?  
+
+Answer:  
```
Once we make the headings themselves questions, the answer can just be the content under that heading and we can get rid of this.

Nadalyn-CZI replied:
I fixed this.

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,115 @@
+* Oftentimes, this comes down to knowing and/or learning the Qt API really well.  
+
+### Better signal or slot documentation or functionality
+Question: How to create better signal/slot documentation/functionality as the user interacts with the core napari interface? For example, creating/editing/deleting a point or shape layer?
```
The question here is suuper vague and tbh I don't really understand what it's asking. I would tweak this to match what we're actually answering.

```suggestion
Question: How can we find the different signals/slots we can connect callbacks to as the user interacts with the core napari interface e.g. creating/editing/deleting a `points` or `shapes` layer?
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,115 @@
+[https://napari.org/guides/stable/events_reference.html](https://napari.org/stable/guides/events_reference.html)  
+Granted, this is a work in progress. 
+
+### Creating or deleting a point or shape layer:  
```
```suggestion
For example, these events are emitted when the user interacts with the layer list:  
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,115 @@
+Granted, this is a work in progress. 
+
+### Creating or deleting a point or shape layer:  
+    Viewer.layers.events.inserted  
```
These should all be in a code block

Nadalyn-CZI replied:
Done.

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,115 @@
+    Viewer.layers.events.changed  
+    Viewer.layers.events.reordered  
+
+Editing a point or shape layer may be complicated, because it may be a pure NumPy object. 
```
```suggestion
Getting an event when the user is editing the data inside a point or shape layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.
```

DragaDoncila reviewed (commented):
Thanks for adding to the TOC, I think this is really close. I've left just some minor suggestions.

The only document that still needs considerable attention imo is the survey. The content is great but the layout is confusing when looking at the built docs: https://output.circle-artifacts.com/output/job/18de8a46-7dcb-4fb2-9fbb-2f51254abba0/artifacts/0/docs/_build/plugins/virtual_environment_docs/5-survey.html#npe2

I've left concrete suggestions on the file.

Nadalyn-CZI committed:
Apply suggestions from code review

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Update 5-survey.md

Removed headings and changed them to questions.

Nadalyn-CZI committed:
Update 5-survey.md

Updated answer about creating and deleting point or shape layer.

Nadalyn-CZI committed:
Update 3-version-management.md

Nadalyn-CZI committed:
Update 5-survey.md

Added code block for list of Viewer.layers.events.

Nadalyn-CZI committed:
Update 2-deploying-your-plugin.md

Adding "pip install build and python -m build ." to demonstrate how to install and run build.

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,114 @@
+
+  **Note:** Gate this action on some criterion, e.g. a git tag as above, or [some other criterion](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).
+
+### Deploying to Anaconda
```
```suggestion
## Deploying to Anaconda
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,102 @@
+## Are there testing environments in napari?
+Answer: Napari does not create or otherwise manage environments.  
+  
+## Introduction to npe2? Migrating to new plug-in architecture?     
```
```suggestion
## Introduction to npe2? Migrating to new plugin architecture?     
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,102 @@
+[https://napari.org/guides/stable/events_reference.html](https://napari.org/stable/guides/events_reference.html)  
+Granted, this is a work in progress. 
+
+## How do you create or delete a point or shape layer?  
```
This is still part of the same question about different signals and slots.

```suggestion
```

Nadalyn-CZI replied:
Removed & incorporated into the signals & slots question.

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
````diff
@@ -0,0 +1,102 @@
+    Viewer.layers.events.reordered  
+```    
+
+Getting an event when the user is editing the data inside a point or shape layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.
````
```suggestion
Getting an event when the user is editing the data inside a `points` or `shapes` layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,102 @@
+* Try not to!
+* You should generally trust that a button click (for example) will call your callback and focus on testing that your callback does what it's supposed to do given that it gets called following some UI interaction.
+* However: If you have a scenario where you are actually creating a complicated widget directly in Qt, see `pytest-qt` for lots of tips, specifically `qtbot`.
+    - pytest-qt - [https://pytest-qt.readthedocs.io/en/latest/intro.html](https://pytest-qt.readthedocs.io/en/latest/intro.html)
```
```suggestion
    - [pytest-qt](https://pytest-qt.readthedocs.io/en/latest/intro.html)
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,102 @@
+* You should generally trust that a button click (for example) will call your callback and focus on testing that your callback does what it's supposed to do given that it gets called following some UI interaction.
+* However: If you have a scenario where you are actually creating a complicated widget directly in Qt, see `pytest-qt` for lots of tips, specifically `qtbot`.
+    - pytest-qt - [https://pytest-qt.readthedocs.io/en/latest/intro.html](https://pytest-qt.readthedocs.io/en/latest/intro.html)
+    - qtbot - [https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)
```
```suggestion
    - [qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,102 @@
+* Tests should not pass/fail together.  
+* Use fixtures to provide a test with inputs, even if you have to make them up.  
+  https://docs.pytest.org/en/6.2.x/fixture.html 
+* Use mocks (mock-ups) to assert that specific calls are made, without necessarily caring about what happens after that call is made.  
```
For both of these the link should be a hyperlink. One on the word `fixtures` and one on the word `mocks`

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,102 @@
+
+Answer:  
+* In single-threaded python, use `Ctrl-C`  
+* In multithreaded python, there are many different patterns. Consider using a generator-based thread worker:  
```
Should be a hyperlink on `generator-based thread worker`

DragaDoncila commented on docs/plugins/virtual_environment_docs/3-version-management.md:
````diff
@@ -0,0 +1,45 @@
+
+   ```console  
+   # configure in pyproject.toml, then…  
+   $ git tag -a v0.1.0 -m v0.1.0```
````
The closing backticks need to be on a new line I believe? The whole thing is rendering as one code block

DragaDoncila reviewed (commented):
This is looking great @Nadalyn-CZI the survey page is looking much better. Final couple of minor comments!

Nadalyn-CZI committed:
Change level of heading.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Change spelling.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Change spelling.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Updated link to pytest-qt.

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Apply suggestions from code review

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Links

Changed fixtures and mocks to links.

Nadalyn-CZI committed:
Removed ## Creating/deleting layers heading. 

Creating/deleting layers was really part of the question about different signals or slots.

Nadalyn-CZI committed:
Added link to generator-based thread worker

Nadalyn-CZI reviewed (commented):
@DragaDoncila, see if this can be merged.  I think it can. 

Thanks for all your help and feedback.

Nadalyn-CZI committed:
Merge branch 'main' into Talley_workshop

Nadalyn-CZI committed:
Corrected code block.

melissawm commented:
Gentle ping @DragaDoncila - anything else we can do here? Thanks!

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,114 @@
+### Building your package 
+`sdist` means source distribution. An `sdist` includes all of the files that are required to *build* your package. An `sdist` may require specific additional software (e.g. compilers) to actually build.
+
+**wheel** is a prebuilt package, ready to drop into site-packages. It includes compiled OS-specific extensions (if applicable).
```
```suggestion
`wheel` is a prebuilt package, ready to drop into your `site-packages` directory. It includes compiled OS-specific extensions (if applicable).
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md:
```diff
@@ -0,0 +1,114 @@
+
+**wheel** is a prebuilt package, ready to drop into site-packages. It includes compiled OS-specific extensions (if applicable).
+
+You are *strongly* encouraged to ship both! If the wheel is not present, pip will try to build it from the sdist before installation, and that may fail depending on the package. To see if a given package ships a wheel, check here: https://pypi.org/project/napari/#files
```
```suggestion
You are *strongly* encouraged to ship both! If the `wheel` is not present, `pip` will try to build it from the `sdist` before installation, and that may fail depending on the package. To see if a given package ships a `wheel`, check here: https://pypi.org/project/napari/#files
```

DragaDoncila commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,98 @@
+* However: If you have a scenario where you are actually creating a complicated widget directly in Qt, see `pytest-qt` for lots of tips, specifically `qtbot`.
+    - [pytest-qt](https://pytest-qt.readthedocs.io/en/latest/intro.html)
+    - [qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)
+* Oftentimes, this comes down to knowing and/or learning the Qt API really well.  
```
We should link to the testing workshop here!

DragaDoncila reviewed (approved):
Just left three final comments but consider this approved from my end! Very excited to have this in :tada:

psobolewskiPhD committed:
Merge branch 'main' into Talley_workshop

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/index.md:
```diff
@@ -0,0 +1,9 @@
+# Tools and tips when building napari plugins or: firehose of stuff I wish I knew earlier  
+
+These guides will be more valuable if you are familiar with the Python programming language and the napari software. They are taken from the [January 2022 Testing workshop](https://www.youtube.com/watch?v=IsHYnI8Tbfw&list=PLilvrWT8aLuYID3YZ7KddS5ky2SaH4DKK) video. These guides stand on their own and are summaries of the information in the video. They are listed here in the order they were present ...
```
Maybe worth clarification that it's the 2nd half of the video by Talley?

psobolewskiPhD commented:
Question/Comment:
the pre-commit action mentioned:
https://github.com/tlambert03/napari-plugin-action
It's archived/not-maintained and npe1 only, not npe2.
So I think we should consider when this info should be removed/deprecated.

psobolewskiPhD committed:
Update docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

psobolewskiPhD committed:
Update docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Nadalyn-CZI committed:
Added link to testing workshop.

Nadalyn-CZI committed:
Just saving my changes.  There is a problem with the link.

Nadalyn-CZI committed:
Update link to testing workshop

Nadalyn-CZI committed:
Update 5-survey.md

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,98 @@
+
+Answer:  
+* In single-threaded python, use `Ctrl-C`  
+* In multithreaded python, there are many different patterns. Consider using a [generator-based thread worker].(https://napari.org/stable/guides/threading.html#generators-for-the-win)    
```
```suggestion
* In multithreaded python, there are many different patterns. Consider using a [generator-based thread worker](https://napari.org/stable/guides/threading.html#generators-for-the-win).    
```

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,98 @@
+* The primary difference is in how plugins are discovered:  
+    - npe1 used decorators, requiring module import.  
+    - npe2 uses static manifests (`napari.yaml`), describing contributions without requiring import.  
+    - Your first plugin: https://napari.org/stable/plugins/first_plugin.html  
```
```suggestion
    - See also the [Your First Plugin tutorial](https://napari.org/stable/plugins/first_plugin.html)  
```

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,98 @@
+  
+* Contributions Reference: https://napari.org/stable/plugins/contributions.html  
+* Guides for each type of contribution: https://napari.org/stable/plugins/guides.html  
+* Migration guide: https://napari.org/stable/plugins/npe2_migration_guide.html  
```
```suggestion
Additional resources:
* [Contributions Reference](https://napari.org/stable/plugins/contributions.html)  
* [Guides for each type of contribution](https://napari.org/stable/plugins/guides.html)  
* [Migration guide](https://napari.org/stable/plugins/npe2_migration_guide.html)  
```

These bare links are clickable in GitHub as markdown, but not in the rendered html pages. This should make them work properly.

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,98 @@
+* Create a script that will start napari and load your widget without any UI interaction.  
+* Don't test as a plugin. Start by directly calling `viewer.window.add_dock_widget` with a manually created widget.  
+* Familiarize yourself with auto-reload features.  
+  https://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html   
```
```suggestion
* Familiarize yourself with the [IPython auto-reload features](https://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html).   
```
Same as above

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,98 @@
+* Familiarize yourself with auto-reload features.  
+  https://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html   
+* Consider using watchmedo from watchdog:  
+  https://github.com/gorakhargosh/watchdog  
```
```suggestion
* Consider using `watchmedo` from [watchdog](https://github.com/gorakhargosh/watchdog).
```

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,98 @@
+    - [pytest-qt](https://pytest-qt.readthedocs.io/en/latest/intro.html)
+    - [qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)
+* Oftentimes, this comes down to knowing and/or learning the Qt API really well.  
+* Please see also the [In-depth guide to plugin testing](../../../../docs/plugins/testing_workshop_docs).
```
```suggestion
* Please see also the [In-depth guide to plugin testing](../testing_workshop_docs/index.md).
```
Need to link to the markdown, I think this should make it work!

psobolewskiPhD committed:
fix link

psobolewskiPhD committed:
make link

psobolewskiPhD committed:
make links

psobolewskiPhD committed:
make link

psobolewskiPhD committed:
make link

psobolewskiPhD committed:
fix link to testing?

psobolewskiPhD commented on docs/plugins/virtual_environment_docs/5-survey.md:
```diff
@@ -0,0 +1,97 @@
+- [How do you avoid github tests failing?](#how-do-you-avoid-github-tests-failing)  
+- [How do you make a process cancellable](#how-do-you-make-a-process-cancellable)  
+- [Are there testing environments in napari?](#are-there-testing-environments-in-napari)  
+- [Introduction to npe2? Migrating to new plug-in architecture?](#introduction-to-npe2-migrating-to-new-plug-in-architecture)  
```
```suggestion
- [Introduction to npe2? Migrating to new plug-in architecture?](#introduction-to-npe2-migrating-to-new-plugin-architecture)  
```

psobolewskiPhD committed:
fix one more link!

psobolewskiPhD commented:
@Nadalyn-CZI I fixed all the links, the key was to link to the `.md` file. And for the others use the proper brackets and parens to make the links clickable.

Going to mark this read-to-merge!

Nadalyn-CZI commented:
God bless you, Peter Sobolewski!  I messed around with it for quite some
time and couldn't get it to work!

Thank you!

On Thu, Jun 8, 2023 at 7:02 AM Peter Sobolewski ***@***.***>
wrote:

> @Nadalyn-CZI <https://github.com/Nadalyn-CZI> I fixed all the links, the
> key was to link to the .md file. And for the others use the proper
> brackets and parens to make the links clickable.
>
> Going to mark this read-to-merge!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/napari/docs/pull/50#issuecomment-1582457573>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3CMNTAPGPFRHR7IJZVNCVLXKG5O3ANCNFSM6AAAAAASCVNCNQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Nadalyn Miller
214-906-5489 personal cell

Nadalyn-CZI commented:
Thank you, Peter Sobolewski!

On Wed, Jun 21, 2023 at 2:26 PM Peter Sobolewski ***@***.***>
wrote:

> Merged #50 <https://github.com/napari/docs/pull/50> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/napari/docs/pull/50#event-9598318520>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3CMNTD3KARM3ETECFRTMVTXMNDG3ANCNFSM6AAAAAASCVNCNQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Nadalyn Miller
214-906-5489 personal cell
